token-metadata/js/src/instruction.ts

```diff
@@ -0,0 +1,53 @@
+import type { PublicKey } from '@solana/web3.js';
+
+/** The field can be one of the required fields (name, symbol, URI), or a
+ * totally new field denoted by a "key" string. */
+type Field = 'name' | 'symbol' | 'uri' | string;
+
+export interface TokenMetadata {
+    // The authority that can sign to update the metadata
+    updateAuthority: PublicKey;
+    // The associated mint, used to counter spoofing to be sure that metadata belongs to a particular mint
+    mint: PublicKey;
+    // The longer name of the token
+    name: string;
+    // The shortened symbol for the token
+    symbol: string;
+    /// The URI pointing to richer metadata
+    uri: string;
+    /// Any additional metadata about the token as key-value pairs
+    additionalMetadata: [string, string][];
```

Comment:
Is this the right schema for the additional properties?

Reply:
This is close! The `updateAuthority` is an optional public key, so typically in Rust we'd represent it with `Option<Pubkey>`, but we're using `OptionalNonZeroPubkey`, which just means for `None` it's going to be a 32-length array of `0`.

So we need to reflect that optional pubkey here somehow. Maybe all we need to do is make the field optional?

Reply:
Can i check if the value will ever be `undefined`?

From what i'm understanding from you (and trying to read the rust code), `updateAuthority` is actually always defined, but could be a 32-length array.

If this is true, while contextually i think it makes sense to make it optional field - if it never actually is optional you force to users to check for undefined (to avoid typescript complaining) when its never true. 

We could create our own type like this below, just to serve as more context
```
// If not defined, will be the default public key with 32-length array of 0.
export type OptionalNonZeroPubkey = PublicKey
```

Reply:
I think making it undefined and then defaulting to the all zeroes is the right move, but the user doesn't really have to know that's happening under the hood, since all they really care about is the fact that the update authority is either the address they've set, or effectively `None`.